src/ir_MilesTag2.cpp

```diff
@@ -0,0 +1,155 @@
+void IRsend::sendMilesShot(const uint64_t data, const uint16_t nbits,
+                      const uint16_t repeat) {
+    _sendMiles(data, nbits, repeat);
+}
+
+/// Send a MilesTag2 formatted MESSAGE packet.
+/// Status: NEEDS TESTING
+/// @param[in] data The message to be sent.
+/// @param[in] nbits The number of bits of message to be sent.
+/// @param[in] repeat The number of times the command is to be repeated.
+void IRsend::sendMilesMsg(const uint64_t data, const uint16_t nbits,
+                      const uint16_t repeat) {
+    _sendMiles(data, nbits, repeat);
+}
+
+void IRsend::_sendMiles(const uint64_t data, const uint16_t nbits,
+                      const uint16_t repeat) {
+  enableIROut(kMilesStdFreq, kMilesStdDuty);
+    // We always send a message, even for repeat=0, hence '<= repeat'.
+  for (uint16_t r = 0; r <= repeat; r++) {
+    // Header
+    mark(kMilesHdrMark);
+    // Data
+    if (nbits == 0)  // If we are asked to send nothing, just return.
+      return;
+    // Send the MSB first.
+    // Send the supplied data.
+    for (uint64_t mask = 1ULL << (nbits - 1); mask; mask >>= 1)
+      if (data & mask) {  // Send a 1
+        space(kMilesSpace);
+        mark(kMilesOneMark);
+      } else {  // Send a 0
+        space(kMilesSpace);
+        mark(kMilesZeroMark);
+      }
+      space(kMilesRptLength);
+  }
+}
+#endif  // SEND_MILESTAG2
+
+#if DECODE_MILESTAG2
+/// Decode the supplied MilesTag2 message.
+/// Status: NEEDS TESTING
+/// @param[in,out] results Ptr to the data to decode & where to store the result
+/// @param[in] offset The starting index to use when attempting to decode the
+///   raw data. Typically/Defaults to kStartOffset.
+/// @param[in] nbits The number of data bits to expect.
+/// @param[in] strict Flag indicating if we should perform strict matching.
+/// @return True if it can decode it, false if it can't.
+/// @see https://github.com/crankyoldgit/IRremoteESP8266/issues/706
+bool IRrecv::decodeMiles(decode_results *results, uint16_t offset,
+                        const uint16_t nbits, const bool strict) {
+  /*
+  uint16_t gap_pos = 0;
+  // we got alot more data than we thought, let's find last GAP and work from it
+  if (results->rawlen >= (2 * nbits + 1)) 
+  {
+      for (uint16_t ind = 0 ; ind < results->rawlen; ind++)
+      {
+          if (matchAtLeast(*(results->rawbuf+ind),34000)) gap_pos = ind;
+      }
+  }
+
+  if (results->rawlen>gap_pos) offset = gap_pos+1;
+  */  
+  // Compliance
+  if (strict) {
+    switch (nbits) {  // Check we've been called with a correct bit size.
+      case 14:
+      case 24:
+        break;
+      default:
+        DPRINT("incorrect nbits:");
+        DPRINTLN(nbits);
+        return false;  // The request doesn't strictly match the protocol defn.
+    }
+  }
+  uint64_t data = 0;
+  // Header
+  if (!matchMark(*(results->rawbuf + offset++),
+                  kMilesHdrMark, kUseDefTol, kMarkExcess)) return 0;
```

Comment:
The document clearly states, that after Header mark 2400us you definetly have a space 600us.
I just decided that it's easier to assume that "first space after header is space of first data bit", Instead of "first space is header space". That way the last bit would have "space and a (zero/one)mark" instead of "(zero/one)mark and silence", since space is 600us silence. Space first then goes a Mark of a bit. That's why i've rewrote genericSend and matchGeneric in my file.